samples/subsys/power/stm32_pm/lowpower/src/main.c

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright (c) 2016 Intel Corporation.
+ * Copyright (c) 2019 STMicroelectronics.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "sample.h"
+
+#define WITHIN_ERROR(var, target, epsilon)       \
+		(((var) >= (target)) && ((var) <= (target) + (epsilon)))
+
+struct device *gpio_port, *gpio_port0, *gpio_port1, *gpio_port2;
+
+struct k_timer lptimer;
+
+/* 1000 msec = 1 sec */
+#define WAIT_TIME 4000
+#define SLEEP_TIME 9000
+
+void button_pressed(struct device *gpiob, struct gpio_callback *cb,
+		    u32_t pins)
+{
+		LL_PWR_ClearFlag_WU();
+		LL_PWR_DisableWakeUpPin(LL_PWR_WAKEUP_PIN2);
+		gpio_pin_write(gpio_port1, LED1, LED_ON);
+		k_busy_wait(100);
+		gpio_pin_write(gpio_port1, LED1, LED_OFF);
+}
+
+static struct gpio_callback gpio_cb;
+
+
+/* Application main Thread */
+void main(void)
+{
+
+	printk("\n\n*** Low Power Demo %s ***\n", CONFIG_BOARD);
+
+	gpio_port = device_get_binding(PORT);
+
+	/* Configure Button to exit lower power mode
+	 * (no Pull up on PA0)
+	 */
+	gpio_pin_configure(gpio_port, SW,
+			   GPIO_DIR_IN | GPIO_INT |
+			   GPIO_INT_EDGE | GPIO_INT_ACTIVE_LOW);
+
+	gpio_init_callback(&gpio_cb, button_pressed, BIT(SW));
+
+	gpio_add_callback(gpio_port, &gpio_cb);
+	gpio_pin_enable_callback(gpio_port, SW);
+
+	/* Configure LEDs */
+	gpio_port0 = device_get_binding(PORT0);
+	gpio_pin_configure(gpio_port0, LED0, GPIO_DIR_OUT);
+	gpio_pin_write(gpio_port0, LED0, LED_ON);
+
+	gpio_port1 = device_get_binding(PORT1);
+	gpio_pin_configure(gpio_port1, LED1, GPIO_DIR_OUT);
+	gpio_pin_write(gpio_port1, LED1, LED_ON);
+
+	gpio_port2 = device_get_binding(PORT2);
+	gpio_pin_configure(gpio_port2, LED2, GPIO_DIR_OUT);
+	gpio_pin_write(gpio_port2, LED2, LED_ON);
+
+	/*
+	 * Start the demo.
+	 */
+
+	while (1) {
+
+		k_sleep(WAIT_TIME);
+
+		printk("<-- Enabling %s state --->\n",
+				       STRINGIFY(SYS_POWER_STATE_SLEEP_1));
+		sys_pm_force_power_state(SYS_POWER_STATE_SLEEP_1);
+
+		k_sleep(SLEEP_TIME);
+
+	}
+}
+
+#ifdef CONFIG_SYS_POWER_SLEEP_STATES
+/*
+ * Application defined function for doing any target specific operations
+ * for low power entry.
+ */
+void sys_pm_notify_power_state_entry(enum power_states state)
+{
+
+	switch (state) {
+
+	case SYS_POWER_STATE_SLEEP_1:
```

Comment:
Each state could be independently enabled/disabled, so you need #if statements around each case.

Reply:
OK